### ISSUE TITLE
Expose FixedTimespan

### ISSUE THREAD

**Issue by ianthetechie** [chronotope/chrono-tz#54]
Hey guys, thanks for making such an awesome crate!

I think it would be quite useful to expose the info currently captured in the private `FixedTimespan` struct. I recognize this may require some collaboration with the chrono crate maintainers, but the general idea would be to either expose the info as a method invocable either on the `DateTime<Tz>` directly or on the `Tz` itself (given a UTC datetime / timestamp) that returns the appropriate `FixedTimespan` so we can get the detailed daylight time or similar offset as well as the display abbreviation at a point in time.

**Comment by quodlibetor** [comment]
Interesting, you specifically want a method that can get you the transition points? What information in particular are you looking for, and what's the use case that you'd like it for?

**Comment by ianthetechie** [comment]
Right now we have the `offset_from_utc_datetime` method that lets you effectively get the absolute UTC offset, but it doesn't include info on 1) whether a daylight (or some other special) time offset is in effect, or 2) the relevant abbreviation to apply to the datetime at this instant (ex: EST vs EDT).

I could get the transition points using some of the underlying crates, but I think a high-level way of exposing the specific offset components (UTC and daylight) as well as the abbreviation at a given point in time makes sense in this crate.

As for the use case, I'm working on API where it would be useful to expose this info to clients. In particular, the abbreviation that's in effect at a given instant, as this can be crossreferenced with the unicode CLDR to come up with a general localized display name (by identifying the correct metazone bazed on the TZ ID and then finding the long equivalent of the abbreviation in the relevant locale file).

I'm not sure how much of this makes sense in chrono_tz directly, but would be happy to contribute if that sounds interesting/useful. At any rate, I think exposing the info from `FixedTimespan` via something like `Tz::offset_at_timestamp` (I'm not sold on my own name; open to suggestions) makes sense.

**Comment by quodlibetor** [comment]
I definitely think that exposing that info makes sense in chrono/chrono-tz!

There are a few things that I think could be done, preferably in a trait so that we can move it into chrono after there's some experience with it:

* `dst() -> chrono::Duration` that returns exactly how offset the time currently is, this is orthogonal to the base offset, I think. I believe that you can get the base offset with `chrono::Offset::fix().local_minus_utc()` or the fields on `FixedTimespan` directly, but I'm not sure if that's what you're referring to here.
* perhaps some function like `is_dst() -> bool` that returns true when savings or political offset is in effect
* Also a convenience method like [Python's `fold` attribute](https://www.python.org/dev/peps/pep-0495) would be useful.

I think that adding a method to chrono_tz::FixedTimespan that provide the abbreviated name is reasonable and can be added to `FixedTimeSpan` directly.

I'd be very happy to take a PR implementing any subset of this stuff, and am happy to discuss!

**Comment by ianthetechie** [comment]
Awesomesauce! Thanks for the quick feedback on this! I'd be happy to write this up and submit a PR, but first want to run a few API design questions by you.

As you said, `chrono::offset::FixedOffset` already directly exposes the *total* UTC offset via `local_minus_utc`, but it hides the DST component. We could just add a new trait (`pub trait DstOffest`?) that exposes this property. However, I could make a design case here for a slightly different trait (not totally sure what to call it... `OffsetComponents`??) that exposes both the base UTC offset and the DST/special offset. The current public interface doesn't really capture that the total offset is a sum of parts, so I think an interface like the following (with docs :D) would provide more clarity.

```rust
pub trait OffsetComponents {
    fn base_utc_offset(&self) -> Duration;
    fn dst_offset(&self) -> Duration;
}
```

Re: `FixedTimespan`, the field for the abbreviation (named `name` at the moment) is already public, but there is no way to access it (the `offset: FixedTimespan` field of `TzOffset` is not public so we can't get the underlying `FixedTimespan` unless I missed something in the API). What are your thoughts on exposing this via a trait `NamedOffset` like this?

```rust
pub trait NamedOffset {
    fn abbreviation(&self) -> &str;
}
```

On a tangentially related note, I'm a bit torn about labeling this as "dst" in the public interface. While almost every additional offset is going to be DST-related, I wonder if a more general name like `special_offset` would be more future-proof.

**Comment by quodlibetor** [comment]
I agree with your position on `OffsetComponents` about the distinction between the base and the special being important. LGTM!

For NamedOffset how would you feel about it looking like:

```
pub trait OffsetName {
    // Guaranteed to exist and to be unambiguous, but may only be a numeric offset e.g. "+0400"
    fn full_name(&self) -> &str;
    // May exist. Multiple timezones may have the same abbreviation
    fn abbreviation(&self) -> Option<&str>;
}
```

I'm actually not sure if it's possible for an abbreviation to not exist in the IANA tzdb, but I suppose the fact that it's not an option in chrono-tz means it probably is guaranteed to exist on the systems that we've ever seen. I think it's plausible that other groups just use the same name for the full name and the abbreviation, so plausibly it would be fine to just return the full name if there is no abbreviation.

> On a tangentially related note, I'm a bit torn about labeling this as "dst" in the public interface.

Yeah, if you can find any other timezone implementation that uses a different name for this I'd be interested -- "dst" feels over-specific but breaking with what folks are used to is more likely to make people annoyed than to cause them to praise us for our foresight.

**Comment by ianthetechie** [comment]
Thanks for the merge! I know you're probably busy, but here's a quick bump to get your thoughts re: 1) `abbreviation` optionality, and 2) full name inclusion / lookup logic.

One other addendum: some time zones are not (currently) part of a metazone (ex: [`Asia/Casablanca`](https://github.com/unicode-org/cldr/blob/master/common/bcp47/timezone.xml#L257)). This presents a slight problem with my approach, as many of these specific time zone names do not appear in [`common/main/<locale>.xml`](https://github.com/unicode-org/cldr/blob/master/common/main/en.xml#L3235). I'm no CLDR expert, so if anyone else could chime in, I'd welcome input on how to best determine the proper "full name" if we want to include it in the trait.

**Comment by quodlibetor** [comment]
Thanks for the ping, I thought I responded to this!

I agree, abbreviation shouldn't be optional.

I was actually thinking of the TZ ID, not the full name, I was using the term wrong. I do agree that getting the full name would be useful. If you wanted to do the work to add it I would be fine putting a bunch of stuff behind a feature gate: vendoring the unicode data files and loading them in at build time, or allowing users to load them dynamically, the unicode license seems to permit that. Currently chrono-tz is pretty focused on compile time stuff, some other lib or a future version could add the features that make it dynamic.

tl;dr:

* Feel free to just add abbreviation, we can bump compatibility of chrono-tz if we add another method
* Adding either `tz_id()` or `full_name()` I think would be great, feel free to add arbitrary dependencies behind a feature gate to make them useful, if you're interested in doing the work.

**Comment by ianthetechie** [comment]
Hey, sorry it took me a few days to get back, and thanks for the thought to this.

I'll add both `tz_id()` and `abbreviation()` to start (PR coming in a moment). After that, I think we can close this issue since now all the useful stuff from `FixedTimespan` is exposed via clean public traits so we can see how this works and maybe promote to chrono for wider use later or w/e.

Re: The full name, I'm still investigating how to handle these "edge cases." I'll look into a few more sources next week to see if there are other ways of doing this. I agree about making this compile time behind a feature gate, vendoring / submoduling the CLDR data, etc. The CLDR version should match the TZDB etc. so I don't personally think a dynamic lookup makes sense, but a feature gate does as this is potentially a lot of extra data. I'll create a separate issue aggregating this discussion next week once I've had a bit of time to sift through some other sources and we can go from there.
